assets/src/components/activities/common/choices/authoring/choiceActions.ts

```diff
@@ -0,0 +1,42 @@
+import { getChoice, getChoices } from 'components/activities/common/choices/authoring/choiceUtils';
+import { Choice, HasChoices, PostUndoable, RichText } from 'components/activities/types';
+
+export const ChoiceActions = {
+  addChoice(choice: Choice) {
+    return (model: HasChoices, post: PostUndoable) => {
+      model.choices.push(choice);
+    };
+  },
+
+  editChoiceContent(id: string, content: RichText) {
+    return (model: HasChoices, post: PostUndoable) => {
+      getChoice(model, id).content = content;
+    };
+  },
+
+  setAllChoices(choices: Choice[]) {
+    return (model: HasChoices, post: PostUndoable) => {
+      model.choices = choices;
+    };
+  },
+
+  removeChoice(id: string) {
+    return (model: HasChoices, post: PostUndoable) => {
+      const choice = getChoice(model, id);
+      const index = getChoices(model).findIndex((c) => c.id === id);
+      model.choices = model.choices.filter((choice) => choice.id !== id);
+
+      post({
```

Comment:
To allow `ChoiceActions` to be reusable in contexts where the removal of a choice also requires the removal of feedback (or execution of other logic), `ChoiceActions.removeChoice` cannot post an undo.  It should instead simply return the choice that has been removed, perhaps also the index that it existed at.  This will then allow the user of `ChoiceActions` to implement and post the correct undoable.

Reply:
This is the problem with using immer without redux backing it, and is one of the reasons I had spent time moving things over to use redux. Immer only lets you mutate a model `draft` or return an entirely new model immutably, but you can't return anything if you mutate the `draft`. So I couldn't find a good way to compose actions when further updates need to know what changes were made.

The trouble with moving to redux was that with the way the models are set up (relational data without being normalized), all of the redux actions to update models were super complicated and depended on the model structure for each activity. The actions were composable, but it was hard to follow the logic when an update made changes to multiple parts of the model.

I played around with a normalization/denormalization step as part of loading/saving, which allowed the actions to be composable and pretty simple, but the normalization had to be run with each keystroke edit, and it forced writing custom `fromPersistence`/`toPersistence` logic for each part of the model.

Bottom line, I'm not sure how to make these actions re-usable. I pushed a change so that the calling actions (in MC and CATA) get the choice and index before they remove the choice, and then use it to post the update. But there's duplication.

Reply:
Elevating the logic to trigger the post of the undo to where you did makes sense.  There is a bit of logic duplication here, I admit, but it keeps the actions more reusable.  The only problem I see is that the undo of removing a choice isn't complete: it also needs to undo removing the corresponding response.